File: objc/Project4/Project4/ViewController.m
//
//  ViewController.m
//  Project4
//
//  Created by Jinwoo Kim on 2/9/21.
//

#import "ViewController.h"

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    if (self.title == nil) return;
    NSURL *url = [NSURL URLWithString:[NSString stringWithFormat:@"https://content.guardianapis.com/%@?api-key=%@&show-fields=thumbnail,headline,standfirst,body", [self.title lowercaseString], apiKey]];
    if (url == nil) return;
    
    dispatch_async(dispatch_get_global_queue(QOS_CLASS_USER_INTERACTIVE, 0), ^{
        [self fetch:url];
    });
}

- (void)fetch:(NSURL *)url {
    // attempt to download the contents of this URL
    NSData *data = [[NSData alloc] initWithContentsOfURL:url];
    if (data) {
        // convert that to JSON and pull out the array we care about
        self.articles = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil][@"response"][@"results"];
        
        // reload the collection view on the main thread
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.collectionView reloadData];
        });
    } else {
        // something went wrong!
    }
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return self.articles.count;
}

- (__kindof UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    NewsCell *newsCell = [collectionView dequeueReusableCellWithReuseIdentifier:@"Cell" forIndexPath:indexPath];
    if (newsCell == nil) {
        [NSException raise:@"NewsCellIsNil" format:@"Unable to dequeue NewsCell from UICollectionView!"];
    }
    
    NSUInteger _idx = 0;
    NSDictionary<NSString *, id> *_newsItem;
    for (NSDictionary<NSString *, id> *dic in self.articles) {
        if (_idx == indexPath.row) {
            _newsItem = dic;
            break;
        }
        _idx += 1;
    }
    NSString *title = _newsItem[@"fields"][@"headline"];
    NSString *thumbnail = _newsItem[@"fields"][@"thumbnail"];
    
    [newsCell.textLabel setText:title];
    
    NSURL *imageURL = [NSURL URLWithString:thumbnail];
    
    if (imageURL) {
        [newsCell.imageView loadURL:imageURL];
    }
    
    return newsCell;
}

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    ReaderViewController *reader = [self.storyboard instantiateViewControllerWithIdentifier:@"Reader"];
    if (reader == nil) return;
    reader.article = self.articles[indexPath.row];
    [self presentViewController:reader animated:YES completion:nil];
}

#pragma UISearchResultsUpdating

- (void)updateSearchResultsForSearchController:(UISearchController *)searchController {
    NSString *text = searchController.searchBar.text;
    if (text == nil) return;
    
    if (text.length == 0) {
        self.articles = @[];
        [self.collectionView reloadData];
    } else {
        NSURL *url = [NSURL URLWithString:[NSString stringWithFormat:@"https://content.guardianapis.com/search?api-key=%@&q=%@&show-fields=thumbnail,headline,standfirst,body", apiKey, [text lowercaseString]]];
        if (url == nil) return;
        dispatch_async(dispatch_get_global_queue(QOS_CLASS_USER_INTERACTIVE, 0), ^{
            [self fetch:url];
        });
    }
}

@end

File: swift/Project4/Project4/ViewController.swift
//
//  ViewController.swift
//  Project4
//
//  Created by Jinwoo Kim on 2/8/21.
//

import UIKit

class ViewController: UICollectionViewController {
    var articles = [JSON]()

    override func viewDidLoad() {
        super.viewDidLoad()
        
        guard let title = title else { return }
        guard let url = URL(string: "https://content.guardianapis.com/\(title.lowercased())?api-key=\(apiKey)&show-fields=thumbnail,headline,standfirst,body")
        else {
            return
        }
        
        DispatchQueue.global(qos: .userInteractive).async {
            self.fetch(url)
        }
    }
    
    func fetch(_ url: URL) {
        // attempt to download the contents of this URL
        if let data = try? Data(contentsOf: url) {
            // convert that to JSON and pull out the array we care about
            articles = JSON(data)["response"]["results"].arrayValue
            
            // reload the collection view on the main thread
            DispatchQueue.main.async {
                self.collectionView?.reloadData()
            }
        } else {
            // something went wrong!
        }
    }

    override func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return articles.count
    }
    
    override func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        guard let newsCell = collectionView.dequeueReusableCell(withReuseIdentifier: "Cell", for: indexPath) as? NewsCell else {
            fatalError("Couldn't dequeue a cell")
        }
        
        let newsItem = articles[indexPath.row]
        let title = newsItem["fields"]["headline"].stringValue
        let thumbnail = newsItem["fields"]["thumbnail"].stringValue
        
        newsCell.textLabel.text = title
        
        if let imageURL = URL(string: thumbnail) {
            newsCell.imageView.load(imageURL)
        }
        
        return newsCell
    }
    
    override func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        guard let reader = storyboard?.instantiateViewController(identifier: "Reader") as? ReaderViewController else {
            return
        }
        reader.article = articles[indexPath.row]
        present(reader, animated: true)
    }
}

extension ViewController: UISearchResultsUpdating {
    func updateSearchResults(for searchController: UISearchController) {
        guard let text = searchController.searchBar.text else { return }
        
        if text.isEmpty {
            articles = [JSON]()
            collectionView?.reloadData()
        } else {
            guard let url = URL(string: "https://content.guardianapis.com/search?api-key=\(apiKey)&q=\(text)&show-fields=thumbnail,headline,standfirst,body") else { return }
            DispatchQueue.global(qos: .userInteractive).async {
                self.fetch(url)
            }
        }
    }
}
